File: frontend/src/local/api/splits.js
/**
 * Splits, local-first — mirrors backend/app/routers/splits.py and
 * split_templates.py so the phone can group and build splits offline.
 *
 * A split is a named group of routines; routine.split_uuid NULL means
 * ungrouped. Deleting a split detaches its routines rather than deleting them
 * — training history must never disappear with a bit of organisation.
 */
import { db, newUuid, nowIso } from '../db'
import { LocalApiError, notFound } from './util'

const stamp = (row) => ({ ...row, updated_at: nowIso(), _dirty: 1 })

// Mirrors SPLIT_TEMPLATES in backend/app/split_templates.py — keep in step.
// [exerciseName, sets]; a name the library doesn't have is skipped.
export const SPLIT_TEMPLATES = {
  ppl: {
    name: 'Push / Pull / Legs',
    days: [
      { name: 'Push', exercises: [['Barbell Bench Press', 4], ['Overhead Press (Barbell)', 3], ['Incline Dumbbell Press', 3], ['Lateral Raise', 3], ['Tricep Pushdown', 3]] },
      { name: 'Pull', exercises: [['Deadlift', 3], ['Pull-Up', 3], ['Seated Cable Row', 3], ['Face Pull', 3], ['Barbell Curl', 3]] },
      { name: 'Legs', exercises: [['Barbell Squat', 4], ['Romanian Deadlift', 3], ['Leg Press', 3], ['Leg Curl (Lying)', 3], ['Standing Calf Raise', 4]] },
    ],
  },
  bro: {
    name: 'Bro Split',
    days: [
      { name: 'Chest', exercises: [['Barbell Bench Press', 4], ['Incline Dumbbell Press', 3], ['Cable Fly', 3], ['Dumbbell Fly', 3]] },
      { name: 'Back', exercises: [['Deadlift', 3], ['Pull-Up', 3], ['Barbell Row', 3], ['Lat Pulldown', 3], ['Seated Cable Row', 3]] },
      { name: 'Shoulders', exercises: [['Overhead Press (Barbell)', 4], ['Lateral Raise', 4], ['Rear Delt Fly', 3], ['Face Pull', 3]] },
      { name: 'Arms', exercises: [['Barbell Curl', 3], ['Incline Dumbbell Curl', 3], ['Skull Crusher', 3], ['Tricep Pushdown', 3], ['Hammer Curl', 3]] },
      { name: 'Legs', exercises: [['Barbell Squat', 4], ['Romanian Deadlift', 3], ['Leg Press', 3], ['Leg Extension', 3], ['Standing Calf Raise', 4]] },
    ],
  },
  legs_glutes: {
    name: 'Legs & Glutes',
    days: [
      { name: 'Legs & Glutes A', exercises: [['Barbell Squat', 4], ['Hip Thrust', 4], ['Romanian Deadlift', 3], ['Leg Curl (Lying)', 3], ['Standing Calf Raise', 4]] },
      { name: 'Legs & Glutes B', exercises: [['Hack Squat', 4], ['Cable Pull-Through', 3], ['Bulgarian Split Squat', 3], ['Leg Extension', 3], ['Seated Calf Raise', 4]] },
    ],
  },
}

async function present(s) {
  const routines = await db.routine.where('split_uuid').equals(s.uuid).toArray()
  return {
    id: s.uuid, name: s.name, position: s.position ?? 0,
    created_at: s.created_at, routine_count: routines.length,
  }
}

async function buildFromTemplate(split, key) {
  const tpl = SPLIT_TEMPLATES[key]
  if (!tpl) throw new LocalApiError(400, `Unknown split template '${key}'`)
  const library = await db.exercise.toArray()
  const byName = new Map(library.map(e => [e.name, e]))

  for (const day of tpl.days) {
    const routine = stamp({
      uuid: newUuid(), name: day.name, source: 'manual', notes: null,
      split_uuid: split.uuid, created_at: nowIso(),
    })
    await db.routine.put(routine)
    let position = 0
    for (const [exerciseName, sets] of day.exercises) {
      const ex = byName.get(exerciseName)
      if (!ex) continue          // trimmed library — skip, don't fail
      await db.routine_exercise.put(stamp({
        uuid: newUuid(), routine_uuid: routine.uuid, exercise_uuid: ex.uuid,
        position, target_sets: sets, target_rep_low: 8, target_rep_high: 12,
        rest_seconds: 120, target_rir: null, planned_sets: null,
      }))
      position += 1
    }
  }
}

async function listSplits() {
  const rows = await db.split.toArray()
  rows.sort((a, b) => (a.position ?? 0) - (b.position ?? 0) || a.name.localeCompare(b.name))
  return Promise.all(rows.map(present))
}

async function createSplit(body) {
  let name = (body.name || '').trim()
  if (!name && body.template) name = SPLIT_TEMPLATES[body.template]?.name || ''
  if (!name) throw new LocalApiError(422, 'A split needs a name')

  const existing = await db.split.toArray()
  const last = existing.reduce((m, s) => Math.max(m, s.position ?? 0), 0)
  const split = stamp({
    uuid: newUuid(), name, position: last + 1, created_at: nowIso(),
  })
  await db.split.put(split)
  if (body.template) await buildFromTemplate(split, body.template)
  return present(split)
}

async function updateSplit(uuid, body) {
  const s = await db.split.get(uuid)
  if (!s) throw notFound('Split')
  const row = stamp({
    ...s,
    name: body.name != null ? body.name.trim() : s.name,
    position: body.position != null ? body.position : s.position,
  })
  await db.split.put(row)
  return present(row)
}

async function deleteSplit(uuid) {
  const s = await db.split.get(uuid)
  if (!s) throw notFound('Split')
  // Detach the days first — deleting a split never deletes routines.
  const routines = await db.routine.where('split_uuid').equals(uuid).toArray()
  for (const r of routines) {
    await db.routine.put(stamp({ ...r, split_uuid: null }))
  }
  await db.split.delete(uuid)
  return null
}

export const splitRoutes = [
  {
    method: 'GET', pattern: /^\/api\/splits\/templates$/,
    handler: () => Object.entries(SPLIT_TEMPLATES).map(([key, t]) => ({
      key, name: t.name, days: t.days.map(d => d.name),
    })),
  },
  { method: 'GET', pattern: /^\/api\/splits$/, handler: () => listSplits() },
  { method: 'POST', pattern: /^\/api\/splits$/, handler: (_m, _q, body) => createSplit(body || {}) },
  { method: 'PUT', pattern: /^\/api\/splits\/([^/]+)$/, handler: (m, _q, body) => updateSplit(m[1], body || {}) },
  { method: 'DELETE', pattern: /^\/api\/splits\/([^/]+)$/, handler: (m) => deleteSplit(m[1]) },
]

File: backend/app/routers/splits.py
"""
Splits — a named group of routines (Push/Pull/Legs, Bro Split, …).

GET    /api/splits        — list this user's splits, with how many days each holds
POST   /api/splits        — create one, optionally built from a template
PUT    /api/splits/{id}   — rename / reorder
DELETE /api/splits/{id}   — delete the split ONLY; its routines survive, ungrouped

Routines carry `split_id`; NULL means ungrouped, which the UI shows as its own
bucket. Deleting a split must never take training history with it, so the
routines are detached explicitly rather than relying on cascade behaviour.
"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func, or_
from sqlalchemy.orm import Session as DBSession

from app.auth import require_auth
from app.db import get_db
from app.models import Exercise, Routine, RoutineExercise, Split, User
from app.schemas import SplitCreate, SplitOut, SplitUpdate
from app.split_templates import SPLIT_TEMPLATES

router = APIRouter(prefix="/api/splits", tags=["splits"])


def _present(db: DBSession, split: Split) -> SplitOut:
    count = (
        db.query(func.count(Routine.id)).filter(Routine.split_id == split.id).scalar() or 0
    )
    return SplitOut(
        id=split.id, name=split.name, position=split.position,
        created_at=split.created_at, routine_count=count,
    )


def _get_owned(db: DBSession, split_id: int, user_id: int) -> Split:
    split = db.get(Split, split_id)
    if not split or split.user_id != user_id:
        raise HTTPException(status_code=404, detail="Split not found")
    return split


def _build_from_template(db: DBSession, user_id: int, split: Split, key: str) -> None:
    """Create the template's days and their exercises under `split`.

    Exercises are referenced by name and resolved against what this user can
    see. An unresolved name is skipped, not fatal — a trimmed library should
    still yield a usable split.
    """
    tpl = SPLIT_TEMPLATES.get(key)
    if tpl is None:
        raise HTTPException(status_code=400, detail=f"Unknown split template '{key}'")

    for day in tpl["days"]:
        routine = Routine(user_id=user_id, name=day["name"], split_id=split.id)
        db.add(routine)
        db.flush()
        position = 0
        for exercise_name, sets in day["exercises"]:
            ex = (
                db.query(Exercise)
                .filter(
                    Exercise.name == exercise_name,
                    or_(Exercise.user_id.is_(None), Exercise.user_id == user_id),
                )
                .first()
            )
            if ex is None:
                continue
            db.add(RoutineExercise(
                user_id=user_id, routine_id=routine.id, exercise_id=ex.id,
                position=position, target_sets=sets,
            ))
            position += 1


@router.get("", response_model=list[SplitOut])
def list_splits(
    db: DBSession = Depends(get_db), current_user: User = Depends(require_auth)
):
    splits = (
        db.query(Split)
        .filter(Split.user_id == current_user.id)
        .order_by(Split.position, Split.name)
        .all()
    )
    return [_present(db, s) for s in splits]


@router.post("", response_model=SplitOut, status_code=201)
def create_split(
    body: SplitCreate,
    db: DBSession = Depends(get_db),
    current_user: User = Depends(require_auth),
):
    if body.template and body.template not in SPLIT_TEMPLATES:
        raise HTTPException(status_code=400, detail=f"Unknown split template '{body.template}'")

    name = (body.name or "").strip()
    if not name and body.template:
        name = SPLIT_TEMPLATES[body.template]["name"]
    if not name:
        raise HTTPException(status_code=422, detail="A split needs a name")

    last = (
        db.query(func.max(Split.position)).filter(Split.user_id == current_user.id).scalar()
    )
    split = Split(user_id=current_user.id, name=name, position=(last or 0) + 1)
    db.add(split)
    db.flush()
    if body.template:
        _build_from_template(db, current_user.id, split, body.template)
    db.commit()
    db.refresh(split)
    return _present(db, split)


@router.put("/{split_id}", response_model=SplitOut)
def update_split(
    split_id: int,
    body: SplitUpdate,
    db: DBSession = Depends(get_db),
    current_user: User = Depends(require_auth),
):
    split = _get_owned(db, split_id, current_user.id)
    if body.name is not None:
        split.name = body.name.strip()
    if body.position is not None:
        split.position = body.position
    db.commit()
    db.refresh(split)
    return _present(db, split)


@router.delete("/{split_id}", status_code=204)
def delete_split(
    split_id: int,
    db: DBSession = Depends(get_db),
    current_user: User = Depends(require_auth),
):
    split = _get_owned(db, split_id, current_user.id)
    # Detach first: the routines (and every session logged against them) stay.
    db.query(Routine).filter(
        Routine.split_id == split.id, Routine.user_id == current_user.id
    ).update({Routine.split_id: None}, synchronize_session=False)
    db.delete(split)
    db.commit()


@router.get("/templates", response_model=list[dict])
def list_templates():
    """The ready-made splits offered when creating one."""
    return [
        {"key": k, "name": t["name"], "days": [d["name"] for d in t["days"]]}
        for k, t in SPLIT_TEMPLATES.items()
    ]
